File: binary_search/binary_search.cpp
#include <vector>
#include <cstdlib>
#include <iostream>
#include <algorithm>
#include "../utility/measure.hpp"
#include "../utility/vectors.hpp"

static int binary_search(std::vector<int>& vec, int val)
{
    int left = 0;
    int right = vec.size()-1;
    int mid = 0;
    while(left <= right)
    {
        mid = (left + right) / 2;
        if (vec[mid] == val)
        {
            return mid;
        }
        if (vec[mid] < val)
        {
            left = mid + 1;            
        }
        else if (vec[mid] > val)
        {
            right = mid - 1;
        }
    }

    return -1;
}

int main(int argc, char* argv[])
{
    int size = 1000;    
    if (argc == 2)
    {
        size = std::strtoul(argv[1], nullptr, 10);
    }
    std::vector<int> v = get_vector(size, 100);    
    std::sort(begin(v), end(v));
    
    int val = random() % 1000;
    perform_print_duration(std::string("Index of ")+std::to_string(val)+" :", binary_search, v, val);        
    
    return 0;       
}

File: binary_search/binary_search.py
def binary_search(arr, elem):
    left = 0
    right = len(arr)-1
    
    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == elem:
            return mid

        if arr[mid] < elem:
            left = mid+1            

        elif arr[mid] > elem:
            right = mid-1            
    return -1

start = 0
end = 7
for i in range(start -2, start + end+4):
    print("-----------------------")
    ndx = binary_search(list(range(start, end, 2)), i)
    if ndx != -1:
        print("{} is present at {}".format(i, ndx))
    else:
        print("{} is not present in the array".format(i))